File: scala-package/examples/src/main/scala/ml/dmlc/mxnet/examples/neuralstyle/end2end/Basic.scala
package ml.dmlc.mxnet.examples.neuralstyle.end2end

import ml.dmlc.mxnet.Shape
import ml.dmlc.mxnet.Context
import ml.dmlc.mxnet.NDArray
import ml.dmlc.mxnet.Symbol
import ml.dmlc.mxnet.Initializer
import org.slf4j.LoggerFactory

/**
 * @author Depeng Liang
 */
object Basic {

  class PretrainedInit(prefix: String, params: Map[String, NDArray],
      verbose: Boolean = false) extends Initializer {

    private val logger = LoggerFactory.getLogger(classOf[PretrainedInit])

    private val prefixLen = prefix.length() + 1
    private val argParams = params.filter(_._1.startsWith("arg:"))
    private val auxParams = params.filter(_._1.startsWith("aux:"))
    private val argNames = argParams.keySet.map(_.substring(4))
    private val auxNames = auxParams.keySet.map(_.substring(4))

    override def initWeight(name: String, arr: NDArray): Unit = {
      val key = name.substring(prefixLen)
      if (this.argNames.contains(key)) {
        if (verbose) logger.info(s"Init $name")
        arr.set(this.argParams(s"arg:$key"))
      } else if (this.auxNames.contains(key)) {
        if (verbose) logger.info(s"Init $name")
        arr.set(this.auxParams(s"aux:$key"))
      } else {
        logger.info(s"Unknown params: $name, init with 0")
        arr.set(0f)
      }
    }

    override def initDefault(name: String, arr: NDArray): Unit = {
    }
  }

  def getStyleModule(prefix: String, dShape: Shape,
      ctx: Context, params: Map[String, NDArray]): Module = {
    val inputShape = Map(s"${prefix}_data" -> dShape)
    val (style, content) = ModelVgg19.getVggSymbol(prefix)
    val (gram, gScale) = styleGramSymbol(inputShape, style)
    val init = new PretrainedInit(prefix, params, true)
    new Module(symbol = gram, context = ctx,
                        dataShapes = Map(s"${prefix}_data" -> dShape),
                        initializer = init, forTraining = false)
  }

  def styleGramSymbol(inputShape: Map[String, Shape], style: Symbol): (Symbol, List[Int]) = {
    val (_, outputShape, _) = style.inferShape(inputShape)
    var gramList = List[Symbol]()
    var gradScale = List[Int]()
    for (i <- 0 until style.listOutputs().length) {
      val shape = outputShape(i)
      val x = Symbol.Reshape()()(Map("data" -> style.get(i),
          "shape" -> Shape(shape(1), shape(2) * shape(3))))
      // use fully connected to quickly do dot(x, x^T)
      val gram = Symbol.FullyConnected()()(Map("data" -> x, "weight" -> x,
          "no_bias" -> true, "num_hidden" -> shape(1)))
      gramList = gramList :+ gram
      gradScale = gradScale :+ (shape(1) * shape(2) * shape(3) * shape(1))
    }
    (Symbol.Group(gramList: _*), gradScale)
  }

  def getLoss(gram: Symbol, content: Symbol): (Symbol, Symbol) = {
    var gramLoss = List[Symbol]()
    for (i <- 0 until gram.listOutputs().length) {
      val gvar = Symbol.Variable(s"target_gram_$i")
      gramLoss = gramLoss :+ Symbol.sum()(Symbol.square()(gvar - gram.get(i))())()
    }
    val cvar = Symbol.Variable("target_content")
    val contentLoss = Symbol.sum()(Symbol.square()(cvar - content)())()
    (Symbol.Group(gramLoss: _*), contentLoss)
  }

  def getContentModule(prefix: String, dShape: Shape,
      ctx: Context, params: Map[String, NDArray]): Module = {
    val (_, sym) = ModelVgg19.getVggSymbol(prefix, true)
    val init = new PretrainedInit(prefix, params)
    new Module(symbol = sym, context = ctx,
                    dataShapes = Map(s"${prefix}_data" -> dShape),
                    initializer = init, forTraining = false)
  }

  def getLossModule(prefix: String, dShape: Shape,
      ctx: Context, params: Map[String, NDArray]): (Module, List[Int]) = {
    val inputShape = Map(s"${prefix}_data" -> dShape)
    val (style, content) = ModelVgg19.getVggSymbol(prefix)
    val (gram, gScale) = styleGramSymbol(inputShape, style)
    val (styleLoss, contentLoss) = getLoss(gram, content)
    val sym = Symbol.Group(styleLoss, contentLoss)
    val init = new PretrainedInit(prefix, params, true)
    val mod = new Module(symbol = sym, context = ctx,
                         dataShapes = Map(s"${prefix}_data" -> dShape),
                         initializer = init, forTraining = true,
                         inputsNeedGrad = true)
    (mod, gScale)
  }
}

File: example/neural-style/end_to_end/basic.py
import sys
sys.path.insert(0, "../../mxnet/python/")

import mxnet as mx
import numpy as np
import model_vgg19 as vgg

class PretrainedInit(mx.init.Initializer):
    def __init__(self, prefix, params, verbose=False):
        self.prefix_len = len(prefix) + 1
        self.verbose = verbose
        self.arg_params = {k : v for k, v in params.items() if k.startswith("arg:")}
        self.aux_params = {k : v for k, v in params.items() if k.startswith("aux:")}
        self.arg_names = set([k[4:] for k in self.arg_params.keys()])
        self.aux_names = set([k[4:] for k in self.aux_params.keys()])

    def __call__(self, name, arr):
        key = name[self.prefix_len:]
        if key in self.arg_names:
            if self.verbose:
                print("Init %s" % name)
            self.arg_params["arg:" + key].copyto(arr)
        elif key in self.aux_params:
            if self.verbose:
                print("Init %s" % name)
            self.aux_params["aux:" + key].copyto(arr)
        else:
            print("Unknown params: %s, init with 0" % name)
            arr[:] = 0.


def style_gram_symbol(input_shape, style):
    _, output_shapes, _ = style.infer_shape(**input_shape)
    gram_list = []
    grad_scale = []
    for i in range(len(style.list_outputs())):
        shape = output_shapes[i]
        x = mx.sym.Reshape(style[i], shape=(int(shape[1]), int(np.prod(shape[2:]))))
        # use fully connected to quickly do dot(x, x^T)
        gram = mx.sym.FullyConnected(x, x, no_bias=True, num_hidden=shape[1])
        gram_list.append(gram)
        grad_scale.append(np.prod(shape[1:]) * shape[1])
    return mx.sym.Group(gram_list), grad_scale


def get_loss(gram, content):
    gram_loss = []
    for i in range(len(gram.list_outputs())):
        gvar = mx.sym.Variable("target_gram_%d" % i)
        gram_loss.append(mx.sym.sum(mx.sym.square(gvar - gram[i])))
    cvar = mx.sym.Variable("target_content")
    content_loss = mx.sym.sum(mx.sym.square(cvar - content))
    return mx.sym.Group(gram_loss), content_loss

def get_content_module(prefix, dshape, ctx, params):
    sym = vgg.get_vgg_symbol(prefix, True)
    init = PretrainedInit(prefix, params)
    mod = mx.mod.Module(symbol=sym,
                        data_names=("%s_data" % prefix,),
                        label_names=None,
                        context=ctx)
    mod.bind(data_shapes=[("%s_data" % prefix, dshape)], for_training=False)
    mod.init_params(init)
    return mod

def get_style_module(prefix, dshape, ctx, params):
    input_shape = {"%s_data" % prefix : dshape}
    style, content = vgg.get_vgg_symbol(prefix)
    gram, gscale = style_gram_symbol(input_shape, style)
    init = PretrainedInit(prefix, params)
    mod = mx.mod.Module(symbol=gram,
                        data_names=("%s_data" % prefix,),
                        label_names=None,
                        context=ctx)
    mod.bind(data_shapes=[("%s_data" % prefix, dshape)], for_training=False)
    mod.init_params(init)
    return mod


def get_loss_module(prefix, dshape, ctx, params):
    input_shape = {"%s_data" % prefix : dshape}
    style, content = vgg.get_vgg_symbol(prefix)
    gram, gscale = style_gram_symbol(input_shape, style)
    style_loss, content_loss = get_loss(gram, content)
    sym = mx.sym.Group([style_loss, content_loss])
    init = PretrainedInit(prefix, params)
    gram_size = len(gram.list_outputs())
    mod = mx.mod.Module(symbol=sym,
                        data_names=("%s_data" % prefix,),
                        label_names=None,
                        context=ctx)
    mod.bind(data_shapes=[("%s_data" % prefix, dshape)],
             for_training=True, inputs_need_grad=True)
    mod.init_params(init)
    return mod, gscale



if __name__ == "__main__":
    from data_processing import PreprocessContentImage, PreprocessStyleImage
    from data_processing import PostprocessImage, SaveImage
    vgg_params = mx.nd.load("./model/vgg19.params")
    style_weight = 2
    content_weight = 10
    long_edge = 384
    content_np = PreprocessContentImage("./input/IMG_4343.jpg", long_edge)
    style_np = PreprocessStyleImage("./input/starry_night.jpg", shape=content_np.shape)
    dshape = content_np.shape
    ctx = mx.gpu()
    # style
    style_mod = get_style_module("style", dshape, ctx, vgg_params)
    style_mod.forward(mx.io.DataBatch([mx.nd.array(style_np)], [0]), is_train=False)
    style_array = [arr.copyto(mx.cpu()) for arr in style_mod.get_outputs()]
    del style_mod
    # content
    content_mod = get_content_module("content", dshape, ctx, vgg_params)
    content_mod.forward(mx.io.DataBatch([mx.nd.array(content_np)], [0]), is_train=False)
    content_array = content_mod.get_outputs()[0].copyto(mx.cpu())
    del content_mod
    # loss
    mod, gscale = get_loss_module("loss", dshape, ctx, vgg_params)
    extra_args = {"target_gram_%d" % i : style_array[i] for i in range(len(style_array))}
    extra_args["target_content"] = content_array
    mod.set_params(extra_args, {}, True, True)
    grad_array = []
    for i in range(len(style_array)):
        grad_array.append(mx.nd.ones((1,), ctx) * (float(style_weight) / gscale[i]))
    grad_array.append(mx.nd.ones((1,), ctx) * (float(content_weight)))
    # train
    img = mx.nd.zeros(content_np.shape, ctx=ctx)
    img[:] = mx.rnd.uniform(-0.1, 0.1, img.shape)
    lr = mx.lr_scheduler.FactorScheduler(step=80, factor=.9)
    optimizer = mx.optimizer.SGD(
            learning_rate = 0.001,
            wd = 0.0005,
            momentum=0.9,
            lr_scheduler = lr)
    optim_state = optimizer.create_state(0, img)

    old_img = img.copyto(ctx)
    clip_norm = 1 * np.prod(img.shape)

    import logging
    for e in range(800):
        mod.forward(mx.io.DataBatch([img], [0]), is_train=True)
        mod.backward(grad_array)
        data_grad = mod.get_input_grads()[0]
        gnorm = mx.nd.norm(data_grad).asscalar()
        if gnorm > clip_norm:
            print("Data Grad: ", gnorm / clip_norm)
            data_grad[:] *= clip_norm / gnorm

        optimizer.update(0, img, data_grad, optim_state)
        new_img = img
        eps = (mx.nd.norm(old_img - new_img) / mx.nd.norm(new_img)).asscalar()
        old_img = new_img.copyto(ctx)
        logging.info('epoch %d, relative change %f', e, eps)
        if (e+1) % 50 == 0:
            SaveImage(new_img.asnumpy(), 'output/tmp_'+str(e+1)+'.jpg')

    SaveImage(new_img.asnumpy(), "./output/out.jpg")

